# Week 02 Assignment glucose level data

Welcome to week two of this course programming 1. You will learn about time related data wrangling with pandas and you will learn to visualize with bokeh. This week will be focussed around missing data. Concretely, you will preprocess the glucose json file using interpolation to impute in order to conduct visual analysis. Learning outcomes:

- load a json dataset 
- typecast the Pandas DataFrame to appropiate data types
- inspect the dataset for quality and metadata information
- add a column with interpolated data in Pandas DataFrame
- perform visual analysis

The assignment consists of 6 parts:

- [part 1: load the data](#0)
     - [Exercise 1.1](#ex-11)
- [part 2: prepare for inspection](#1)
     - [Exercise 2.1](#ex-21)
- [part 3: inspect the data](#2)
     - [Exercise 3.1](#ex-31)
- [part 4: interpolate the data](#3)
     - [Exercise 4.1](#ex-41)
- [part 5: visualize the data](#4)
     - [Exercise 5.1](#ex-51)
- [part 6: Challenge](#5)
     - [Exercise 6.1](#ex-61)

Part 1 and 5 are mandatory, part 6 is optional (bonus)
To pass the assingnment you need to a score of 60%. 


<a name='0'></a>
## Part 1: Load the data

Instructions: Load the json datafile `glucose.json` into a pandas dataframe. Check your dataframe with a `.head()` to compare with the expected outcome

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>json.load() method reads a file, pd.read_json converts it to a Pandas DataFrame</li>
    <li>when loading into a Pandas DataFrame use records orientation </li>
</ul>
</details>

<a name='ex-11'></a>
### Code your solution

In [43]:
# CODE YOUR SOLUTION HERE
import pandas as pd
import numpy as np
import json as js
from bokeh.io import output_notebook
output_notebook()
f = open('D:\programming1\BFVM19PROG1\data\glucose.json')
df= js.load(f)
df=pd.read_json(df)
df
#df.head()



Loading BokehJS ...

ID              time  recordtype glucose
0    2.845000e+03  2019-04-25 00:08           1     109
1    2.850000e+03  2019-04-25 00:50           1        
2    2.877000e+03  2019-04-25 07:02           1     123
3    2.881000e+03  2019-04-25 07:34           1     158
4    2.886000e+03  2019-04-25 08:19           1        
..            ...               ...         ...     ...
131  1.614305e+19  2019-04-25 22:47           0    None
132  1.614305e+19  2019-04-25 23:02           0    None
133  1.614305e+19  2019-04-25 23:18           0    None
134  1.614305e+19  2019-04-25 23:33           0    None
135  1.614305e+19  2019-04-25 23:48           0    None

[136 rows x 4 columns]

#### Expected outcome: 

<a name='1'></a>
## Part 2: Prepare the data

Check the datatypes of your dataframe. The `glucose` field should be an integer, the `time` field should have a datetime format. If the datatypes are different you should typecast them to the right format.
Make sure that your dataset is sorted by the time column


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use astype() method or pandas.DataFrame.to_datetime() for instance</li>
    <li>make sure that the empty spaces are filled with NaN. Use errors='coerce'</li>
    <li>set_index(), sort_index() and reset_index() are helpful to sort on index</li>
</ul>
</details>

<a name='ex-21'></a>
### Code your solution

In [44]:
# CODE YOUR SOLUTION HERE
df['glucose']=pd.to_numeric(df['glucose'], errors='coerce',downcast='signed')
df.dtypes
#df['glucose']=df['glucose'].astype(pd.Int64Dtype())
df['time']=pd.to_datetime(df['time'])
df['ID']=df['ID'].astype(np.int64)
print(df.dtypes)
df=df.set_index(['time'])
df=df.sort_index()




ID                     int64
time          datetime64[ns]
recordtype             int64
glucose              float64
dtype: object


#### Expected outcome: 

<a name='2'></a>
## Part 3: Inspect the data

Now that we prepared the data we are going to inspect the data to get more familiar with the data. You are required to do the following

- inspect the percentage missing data for glucose
- what is the relationship between recordtype and glucose value?
- what is the relationship between ID and glucose value?

Code the solutions to your answers. Create meaningful overviews or statistics

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>In the week 01 assignment some functions were explained to inspect missing values</li>
    <li>In the week 01 assignment some functions were explained to groupby value</li>
</ul>
</details>

<a name='ex-31'></a>
### Code your solution

In [45]:
#CODE YOUR SOLUTION HERE
miss_value=(df.isna().sum()/len(df))
print(miss_value[2])

print(df['glucose'].corr(df['recordtype']))
print(df['glucose'].corr(df['ID']))



0.6176470588235294
nan
-0.20610584631462361


#### Expected outcome percentage missing data
0.6176470588235294

<a name='3'></a>
## Part 4: Interpolate the data

A lot of data is missing. Use interpolation to fill the missing values. Create a new column with the interpolated data. Take an argumentative approach. Select an interpolation method that suits the nature of the data and explain your choice. Mind you that the expected outcome of the interpolation values can differ from the example below

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use Pandas.DataFrame.interpolate() method</li>
</ul>
</details>

<a name='ex-41'></a>
### Code your solution

In [46]:
#CODE YOUR SOLUTION HERE
df['interpolated']=df['glucose']
df1=df.copy()
df['interpolated'].interpolate(method='time',limit_direction='forward', inplace=True)
df.head()



ID  recordtype  glucose  interpolated
time                                                                       
2019-04-25 00:08:00                 2845           1    109.0    109.000000
2019-04-25 00:14:00 -9223372036854775808           0      NaN    109.202899
2019-04-25 00:29:00 -9223372036854775808           0      NaN    109.710145
2019-04-25 00:44:00 -9223372036854775808           0      NaN    110.217391
2019-04-25 00:50:00                 2850           1      NaN    110.420290

#### Example outcome

<a name='4'></a>
## Part 5: Plot the data

Create a plot with the original data and the interpolated data. Consider what the best representation is for visualisation of actual values and modelled/imputed values. An example of such a plot is given below. This plot however is not considered the best practice. 

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>figure(x_axis_type='datetime') automatically makes nices labels of the datetime data</li>
</ul>
</details>

<a name='ex-51'></a>
### Code your solution

In [47]:
from bokeh.io import output_notebook
from bokeh.plotting import figure, show
from bokeh.layouts import gridplot
from bokeh.io import output_notebook
from bokeh.plotting import ColumnDataSource
from bokeh.models.formatters import DatetimeTickFormatter
output_notebook()

Loading BokehJS ...

In [48]:
#CODE YOUR SOLUTION HERE
ColumnDataSource=df.copy()

fig=figure(title="plot for interpolated with time method",x_axis_label='time',plot_width = 500, plot_height = 400,x_axis_type='datetime')
fig.line(x=ColumnDataSource.index,y=ColumnDataSource['interpolated'], legend="Temp.", line_width=2, color='blue')
fig.line(x=ColumnDataSource.index,y=ColumnDataSource['glucose'], legend="Temp.", line_width=2, color='red')
fig.xaxis.formatter = DatetimeTickFormatter(days="%d-%b-%Y", hours="%H:%M", seconds="%S" )
show(fig)


<a name='6'></a>
## Part 6: Challenge

It might even be interesting to introduce a widget in which you can select different methods to interpolate.
1. Can you improve the interpolation by choosing an other method?
2. Can you add an rolling mean line? 
2. Can you improve the plot by making it interactive?

<a name='ex-61'></a>
### Code your solution

In [49]:
#CODE YOUR SOLUTION HERE
def show_fig(ColumnDataSource,method_name,color):
    source=ColumnDataSource.copy()
    source['interpolated'].interpolate(method=method_name,limit_direction='forward', inplace=True)
    fig=figure(title="plot for interpolated with "+method_name+" method",x_axis_label='time',y_axis_label='interpolated data',x_axis_type='datetime',\
    plot_width = 300, plot_height = 300)
    fig.line(x=source.index,y=source['interpolated'], legend="Temp.", line_width=2,color=color)
    fig.xaxis.formatter = DatetimeTickFormatter(days="%d-%b-%Y", hours="%H:%M", seconds="%S" )
    return fig
fig1=show_fig(df1,'pad','orange')
fig2=show_fig(df1,'time','pink')
fig3=show_fig(df1,'linear','green')
grid = gridplot(children = [[fig1,fig2,fig3]],sizing_mode='fixed',plot_width=450, plot_height=400)
show(grid)
df.rolling(4,center=True).sum()


    



ID  recordtype  glucose  interpolated
time                                                                
2019-04-25 00:08:00           NaN         NaN      NaN           NaN
2019-04-25 00:14:00           NaN         NaN      NaN           NaN
2019-04-25 00:29:00 -2.767012e+19         1.0      NaN    438.130435
2019-04-25 00:44:00 -2.767012e+19         1.0      NaN    439.550725
2019-04-25 00:50:00 -2.767012e+19         1.0      NaN    441.072464
...                           ...         ...      ...           ...
2019-04-25 23:02:00 -2.767012e+19         1.0      NaN    431.035714
2019-04-25 23:18:00 -1.844674e+19         2.0      NaN    433.750000
2019-04-25 23:31:00 -2.767012e+19         1.0      NaN    438.750000
2019-04-25 23:33:00 -2.767012e+19         1.0      NaN    442.375000
2019-04-25 23:48:00           NaN         NaN      NaN           NaN

[136 rows x 4 columns]